### DISCUSSION TITLE
Support for the Maestro testing tool - `@nx/maestro`

### DISCUSSION THREAD

**Discussion by tmc** [nrwl/nx#19473] [Feature Requests]
## Description
[Maestro](https://maestro.mobile.dev/) is an emerging tool in the native testing space, it'd be great to have `@nx/maestro` which would provide equivalents to https://nx.dev/nx-api/detox but leverage the Maestro tool.
 
## Suggested Implementation
I imagine copying capabilities from detox that make sense would be the move.

**Comment by AgentEnder** [comment]
This would be a good candidate for a community plugin, but isn't something that we currently use or have a way to dogfood ourselves so we aren't likely to consider it at the time being. For now, I'm going to close this issue and convert it to a discussion. If more folks are interested or we start to see trends towards it over detox we could consider it in the future.

**Reply by leggomuhgreggo** [reply]
FWIW it does seem like Expo is aligning on Maestro. Still early days — but it may be something to consider in the long term.

**Reply by hassankhan** [reply]
Detox _frequently_ lags behind on the latest React Native version; it currently supports up to [React Native v0.77](https://github.com/wix/Detox/blob/c8724e60a59a47c80c7424a2f5cb54812cfa5f9f/README.md?plain=1#L63), and while it might work with newer versions, it is most definitely not guaranteed due to the way its designed.

Aside from newer version support being slow, Detox requiring a specialized build step before being able to run tests is also a significant time sink. Maestro on the other hand runs against the same Android/iOS binary that's created with either the Expo or React Native CLI.

**Reply by leggomuhgreggo** [reply]
Wanted to add, by way of follow-up, that:

- As of recently, Expo has fully dropped support for Detox.
- Maestro _does_ seems to be emerging as the new standard for native testing

_Also FWIW — while I was skeptical at first — having used both, I do think Maestro is an improvement_

Was thinking aboout what would be involved, if we forked the detox executor and adapted it to maestro. Added some notes:

<details>

  <summary>Notes</summary>

- **executors:** test, build
- **generators:** init, application, convert-to-inferred
- **misc:** plugin files; companion support in Nx plugins for expo + react-native; dependency versions; cache + outputs; default config; etc.
- **key differences:**
  - maestro installed as system-level dependency (eg homebrew etc), as opposed to npm package
  - maestro doesn't use jest
  - maestro is configured with yaml files 
  
  Related Docs:
  - https://docs.maestro.dev/getting-started/installing-maestro
  - https://docs.expo.dev/eas/workflows/reference/e2e-tests/
  - https://docs.swmansion.com/setup-ci/docs/workflows/maestro/
  
</details>

**Reply by hassankhan** [reply]
Maestro projects are conventionally placed within a project under `.maestro/` (similar to Storybook), so an Nx integration would probably want to set it up under a separate `-e2e` project, as it does with Detox.

**Reply by leggomuhgreggo** [reply]
FWIW: +1 for separate project (at least as default) — there are tradeoffs, but it makes it easier to support E2E CI pattens for "consumer driven API tests" via `implicitDependencies` (ie if api changes, run e2e, but not app build/deploy) — also consistency with existing Nx E2E pattens.

**Reply by corners99** [reply]
Let's get this happening
